src/realm/realmReducer.js

```diff
@@ -24,6 +24,7 @@ const initialState = {
   twentyFourHourTime: false,
   canCreateStreams: true,
   isAdmin: false,
+  isTopicMandatory: false,
```

Comment:
Let's add a `dropCache` migration in src/boot/store.js for this. Something like,

```js
  // Add `isTopicMandatory` to `state.realm`.
  '30': dropCache,
```

(or a higher number than 30, if new migrations happen to get added after I post this review). See `dropCache`'s jsdoc for what it's for.

Reply:
Does this actually matter? My understanding is that an app update will drop the cache anyways, since it'll force a restart, and we aren't persisting this data in any case. I could be misunderstanding how long we intend for this data to persist, though.

Reply:
Ah: we are persisting this data. `realmReducer` deals with `state.realm`, and 'realm' is in `cacheKeys`. Here's the jsdoc on `cacheKeys`:

```js
/**
 * Properties on the global state which we persist for caching's sake.
 *
 * These represent information for which the ground truth is on the
 * server, but which we persist locally so that we have it cached and
 * don't have to re-download it.
 */
```

Reply:
That data won't get cleared automatically on an app restart, but it's easy and harmless for us to tell it to clear on a version upgrade, by using a `dropCache` migration.

Reply:
Hmm, but since this is getting overwritten on the `/register` response, doesn't that mean that it's going to get overwritten on app load anyways, since we're going to call `/register` then, before we display anything?

(To be clear, I think it's reasonable to have a `dropCache` in either case, I'm just trying to make sure I have a good understanding of the data flow)

Reply:
> (To be clear, I think it's reasonable to have a `dropCache` in either case, I'm just trying to make sure I have a good understanding of the data flow)

Makes sense! 🙂

We do expect this to get clobbered with an up-to-date value when a `/register` completes. But there are two important possible states to consider while that `/register` is in progress:

- We have no cached server data to show, so we show `FullScreenLoading`.
- We have some stale data for this account, cached from a previous session, so we show an interactable UI with the "Connecting..." banner (`LoadingBanner`).

The conditional for what to show is provided by `useHaveServerDataGate`.

(The logic for what to show used to be more complicated and pretty hard to verify, depending on navigating imperatively at the right time, and setting the initial nav state correctly. If you want, you can read through #4454—particularly, a023b7540—for where we looked at that logic closely and refactored it by removing 'loading' as a navigation route; we then consolidated a reusable "have-server-data gate" in #4603 for use with components besides `MainTabsScreen`, and we fixed a bug in that "gate" logic in 98d19b829 just a few weeks ago.)

In that second state—

> - We have some stale data for this account, cached from a previous session, so we show an interactable UI with the "Connecting..." banner.

—it's possible, in theory (and I guess in practice, if the `/register` takes super long and we fix various UI perf issues) to go poking around the app and try to do things with the stale data that you'd also want to do with up-to-date data. Like, I guess, try to send a message without a topic. So it's good to do something explicit like this to ensure we have the right data from the beginning.

Reply:
> So it's good to do something explicit like this to ensure we have the right data from the beginning.

Hmm, even more explicit than `dropCache` would be if the migration were written to add the key explicitly. I think `dropCache` would work, but only because of the "shallow-merge" quirk in `redux-persist` I found over at https://github.com/zulip/zulip-mobile/pull/4679#discussion_r626022077. (And that same quirk may have the effect that things would work even if we omitted a migration—but as I said there, it's nice to be as explicit as possible with migrations.)

I ~copied the `dropCache` idea from https://github.com/zulip/zulip-mobile/pull/4543#discussion_r627462602.

So, now I'm thinking something like

```js
  // Add `isTopicMandatory` to `state.realm`.
  '30': state => ({
    ...state,
    realm: {
      ...state.realm,
      isTopicMandatory: false,
    },
  }),
```

Sorry for the confusion!

Reply:
> And that same ["shallow-merge"] quirk may have the effect that things would work even if we omitted a migration altogether

Yes, that's right. Effectively it means that when we introduce a new top-level property to any of our top-level state subtrees -- anything of the form `state.foo.bar` -- then without doing anything about it in a migration, we get the same effect as if we had a migration to give it the same value that we have for it in the initial state, as specified in the reducer.

A key bit of the background to how it works out that way is that the top-level subtrees `state.foo` are the units of what we store with `redux-persist`.

> I think `dropCache` would work, but only because of the "shallow-merge" quirk

The reason `dropCache` works is similar in spirit but I think it's more basic: it's that if a given top-level subtree is missing entirely, then at rehydrate time we just let the initial state (as specified in the reducer) stand untouched.

---

So in a situation like this, where we've introduced a new property `state.foo.bar` on one of the top-level state subtrees `state.foo`, the result is that any of:
1. no migration
2. a `dropCache` migration
3. a migration that explicitly adds the new property, like in Chris's last comment above

will produce the same, correct results.

It's good to be explicit, as with option 3. On the other hand, it means we duplicate the detail of what the default is -- new installs will get the default from the reducer's initial state, while old installs upgrading will get it from the migration. It also means more boilerplate, in a spot that isn't really type-checked (and is fundamentally tricky to properly type-check) and where getting it wrong could be pretty unpleasant.

So I'm kind of inclined to actually make our practice for this situation be option 1: no migration at all.

However! If we do that, I think it's important to still be explicit about making that choice. That helps us be in the habit of thinking about migrations when updating our types, so that we explicitly think through whether the update is one that falls into this situation and doesn't need a migration. So that could look like adding a comment here, the same place we'd add a migration, that looks like:
```js

  // Add `isTopicMandatory` to `state.realm`.
  // No migration; just use default from initial state.
```

---

Scanning back through previous migrations, I think the only one to date that we'd actually have skipped under this approach is 28, adding `state.settings.browser`. All the others either do something more complicated than adding a property, or add a property somewhere deeper in the tree (like in each element of `state.accounts`.)

As we add more things like miscellaneous settings, though, this could be something that becomes more common.

Reply:
Made a chat thread to help make this discussion easier to find: https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/data.20migrations/near/1230029